### PR TITLE
fix: Step order is broken

### PR DESCRIPTION
**Pull request by ibu1224**
## Context

I found a build where the order of the steps was broken.
It happens several times a day.

image

The expected steps order are: `1 2 3 .... 5 6 7`.
The cause probably occurs when `stepFactory.create` is delayed.

## Objective
In this PR, modify to wait for `stepFactory.create`.

## References

https://stackoverflow.com/questions/41243468/javascript-array-reduce-with-async-await

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/buildFactory.js [hunk 1]
```diff
@@ -263,8 +263,9 @@ class BuildFactory extends BaseFactory {
 
                     const build = await super.create(modelConfig);
 
-                    await steps.reduce((p, step) =>
-                        p.then(() => stepFactory.create(Object.assign(
+                    await steps.reduce(async (p, step) =>
+                        // eslint-disable-next-line no-return-await
+                        await p.then(() => stepFactory.create(Object.assign(
```
Comment (kumada626) [inline lib/buildFactory.js]: Don't we have to await `stepFactory.create() `?

Comment (jithine) [reply lib/buildFactory.js]: perhaps switch to `for of` for ensuring order. We do something similar here https://github.com/screwdriver-cd/screwdriver/pull/2011/files#diff-7a514e440aa937beee722c7ac84dfb93R1263-R1271

Comment (ibu1224) [reply lib/buildFactory.js]: @jithine 
It is a similar case.
Modify it using the `for of`.

File: lib/buildFactory.js [hunk 2]
```diff
@@ -268,10 +268,16 @@ class BuildFactory extends BaseFactory {
...
-                        Promise.resolve()
-                    );
+                    // eslint-disable-next-line no-restricted-syntax
+                    for (const step of steps) {
+                        try {
+                            // eslint-disable-next-line no-await-in-loop
+                            await stepFactory.create({ buildId: build.id, ...step });
+                        } catch (err) {
+                            // eslint-disable-next-line max-len
+                            throw new Error(`Error in steps -BuildId:${build.id}-step:${step}`, err);
```
Comment (kumada626) [inline lib/buildFactory.js]: I think it is also better to log these message using `logger.error`.

Comment (ibu1224) [reply lib/buildFactory.js]: ok. I modify to using `logger`

**Review by wahapo — APPROVED** [review]

**Review by yoshwata — APPROVED** [review]

File: lib/buildFactory.js [hunk 3]
```diff
@@ -268,10 +269,15 @@ class BuildFactory extends BaseFactory {
...
-                        (p, step) => p.then(() => stepFactory.create({ buildId: build.id, ...step })),
-                        Promise.resolve()
-                    );
+                    // eslint-disable-next-line no-restricted-syntax
+                    for (const step of steps) {
+                        try {
+                            // eslint-disable-next-line no-await-in-loop
+                            await stepFactory.create({ buildId: build.id, ...step });
+                        } catch (err) {
+                            logger.error(`Error in steps -BuildId:${build.id}-step:${step}`, err);
```
Comment (tkyi) [inline lib/buildFactory.js]: Could have error message something like below to be more consistent with https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/builds/index.js#L1346-L1347, but not a big deal if we leave it as-is

```suggestion
                            logger.error(`Error in buildFactory create steps - buildId:${build.id}-step:${step}`, err);
```

**Review by tkyi — APPROVED** [review]
LGTM
small suggestion

File: lib/buildFactory.js (hunk 3 above)
Comment (ibu1224) [reply lib/buildFactory.js]: Thank you review.
I modify error message.

File: lib/buildFactory.js [hunk 4]
```diff
@@ -268,10 +269,15 @@ class BuildFactory extends BaseFactory {
...
-                        (p, step) => p.then(() => stepFactory.create({ buildId: build.id, ...step })),
-                        Promise.resolve()
-                    );
+                    // eslint-disable-next-line no-restricted-syntax
+                    for (const step of steps) {
+                        try {
+                            // eslint-disable-next-line no-await-in-loop
+                            await stepFactory.create({ buildId: build.id, ...step });
+                        } catch (err) {
+                            logger.error(`Error in BuildFactory create -buildId:${build.id}-step:${step}`, err);
```
Comment (jithine) [inline lib/buildFactory.js]: can you add a space after first `-` ?
```Error in BuildFactory create - buildId```

Comment (ibu1224) [reply lib/buildFactory.js]: OK. I add space in front of `buildId`.

**Review by kumada626 — APPROVED** [review]

**Pull request merged**